### ISSUE TITLE
Reorganize local cache

### ISSUE THREAD

**Issue by candleindark** [datalad/datalad-registry#125]
Currently, the initial layer of the cache are directories based on UUID of datasets. Consider naming the directories using the first few letters of the hashed representation of URLs instead.

**Comment by yarikoptic** [comment]
I also wonder if URL could be represented in some better form, still reflective of original URL, than base64?

ha -- we should adopt one used by git-annex itself for URL backend

```
❯ git annex addurl --fast https://github.com/datalad/datalad-registry/issues/125
addurl https://github.com/datalad/datalad-registry/issues/125 (to github.com_datalad_datalad-registry_issues_125) ok
(recording state in git...)
❯ ls -l
lrwxrwxrwx 1 yoh yoh 146 Mar 10 13:10 github.com_datalad_datalad-registry_issues_125 -> .git/annex/objects/50/Wg/URL--https&c%%github.com%datalad%datalad-registry%issues%125/URL--https&c%%github.com%datalad%datalad-registry%issues%125
```
so pretty much it is just URL encoding of the url itself!

```shell
In [13]: from urllib.parse import quote

In [14]: quote('https://github.com/datalad/datalad-registry/issues/125', safe='')
Out[14]: 'https%3A%2F%2Fgithub.com%2Fdatalad%2Fdatalad-registry%2Fissues%2F125'
```

OR even better -- why not to use use the

```
In [15]: quote('https://github.com/datalad/datalad-registry/issues/125')
Out[15]: 'https%3A//github.com/datalad/datalad-registry/issues/125'
```
 and without any leading folders based on checksum since we would get hierarchy naturally now unless we have thousands of subfolders under 1 single domain path, besides like for dandizarrs etc. But overall I don't think it is often to happen, so I would not even sweat about it.


Here is a list of all in usage dashboard

```shell
$ jq '.github[].url' < datalad-repos.json | tr '"' ' ' | python3 -c 'from urllib.parse import quote; import sys; print("\n".join(quote(l.strip()) for l in  sys.stdin.readlines()))'
```
available at http://www.oneukrainian.com/tmp/datalad-repos-urlquoted.txt .

I didn't check though either at the level current API we would not get a trouble using such URLs with `/` in them. If we do -- may be within API requests they would be needed to be "without safe /", so we would do `unquote/quote(safe="/")` to get to the local folder one.

**Comment by candleindark** [comment]
I prefer to have the URLs to be encoded in a human readable form as well.

When you mentioned
```
In [15]: quote('https://github.com/datalad/datalad-registry/issues/125')
Out[15]: 'https%3A//github.com/datalad/datalad-registry/issues/125'
```

Do you mean to have a directory `github.com`, then a subdirectory `datalad` underneath it and so on?

**Comment by candleindark** [comment]
I am not concern about the compatibility of the URL encoding with the API. The main function that receives dataset URLs from a client is https://github.com/datalad/datalad-registry/blob/master/datalad_registry/datasets.py#L41-L77. I think it should be changed in the future so that it takes dataset URLs from a web form or in JSON. In that case, we don't have to worry about dataset URL encoding for the web API at all. Additionally, we a client can pass multiple dataset URLs to the server at one time.

**Comment by yarikoptic** [comment]
> Do you mean to have a directory `github.com`, then a subdirectory `datalad` underneath it and so on?

Yes. FWIW it does add a bit of burden on clean up to possibly remove empty parent directories, which might be leading to race conditions. So may be adding that extra layer to just make it statistically improbable would be worthwhile after all

**Comment by candleindark** [comment]
> So may be adding that extra layer to just make it statistically improbable would be worthwhile after all

Agree. Using hashed names in the extra layer will also ensure the datasets to be more evenly distributed into directories in this top extra layer.

**Comment by candleindark** [comment]
Can we decide on having the first layer as the first 3 characters of the hash or the URL, and the second layer as

```
In [13]: from urllib.parse import quote

In [14]: quote('https://github.com/datalad/datalad-registry/issues/125', safe='')
Out[14]: 'https%3A%2F%2Fgithub.com%2Fdatalad%2Fdatalad-registry%2Fissues%2F125'
```

for readability?

**Comment by yarikoptic** [comment]
fine with me.

**Comment by candleindark** [comment]
> ```
> In [13]: from urllib.parse import quote
> 
> In [14]: quote('https://github.com/datalad/datalad-registry/issues/125', safe='')
> Out[14]: 'https%3A%2F%2Fgithub.com%2Fdatalad%2Fdatalad-registry%2Fissues%2F125'
> ```

I actually have additional concern about organization of the second layer. Will this arrangement be a problem if we encounter a very long URL?

**Comment by yarikoptic** [comment]
rright - good point, and I bet it was one of the concerns why we did not do that way. on some systems it might be of concern. In `datalad wtf` we test current path for maximal file path IIRC, e.g. on typhon
```
yoh@typhon:/tmp$ datalad wtf -S system
# WTF
## system 
  - distribution: debian/n/a/bookworm
  - encoding: 
    - default: utf-8
    - filesystem: utf-8
    - locale.prefered: UTF-8
  - filesystem: 
    - CWD: 
      - max_pathlength: 4096
      - mount_opts: rw,relatime
      - path: /tmp
      - type: ext4
    - HOME: 
      - max_pathlength: 4096
      - mount_opts: rw,noatime,compress=lzo,ssd,discard=async,space_cache=v2,subvolid=256,subvol=/home
      - path: /home/yoh
      - type: btrfs
    - TMP: 
      - max_pathlength: 4096
      - mount_opts: rw,relatime
      - path: /tmp
      - type: ext4
  - max_path_length: 260
  - name: Linux
  - release: 6.1.0-3-amd64
  - type: posix
  - version: #1 SMP PREEMPT_DYNAMIC Debian 6.1.8-1 (2023-01-29)
```

so even though according to psutil info about max path length to be 4096,  in this folder on `/tmp` I am at ATM I could only create filename of length `260 - len("/tmp")` and no longer:

```
yoh@typhon:/tmp$ touch `python -c 'print("a"*256)'`
touch: cannot touch 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa': File name too long
yoh@typhon:/tmp$ touch `python -c 'print("a"*255)'`
```

so indeed we can't assume that we would not hit such a limit.
Note: I never plan to deploy on Windows where I believe we would immediately run into that range of limitations in the total path length.  One, still feasible solution is to do use subfolders

```
In [15]: quote('https://github.com/datalad/datalad-registry/issues/125')
Out[15]: 'https%3A//github.com/datalad/datalad-registry/issues/125'
```
It is not ideal but I think pragmatically it should accommodate all URLs, BUT we need to think more
- might not be good enough when we start hitting those new ones for `datalad-annex` git-remote as introduced in https://github.com/datalad/datalad-next/blob/HEAD/datalad_next/gitremotes/datalad_annex.py#L53 . There `?options` in URL could be quite long
- http://example.com/ could be legit and http://example.com/someother too. (although typically they would terminate with `.git/` but it doesn't have to be that way).  So we should terminate them with something which also would unlikely to happen (not allowed) in url quoted URL, e.g. the `?` itself as there is no options.  All the `?options` and `#fragments` we can base64 encode thus hope for shorter representation.  So what about smth like

```python
urlonly,options=url.split('?', 1) if '?' in url else url, ""
path = '{quote(urlonly)}/?{base64.urlsafe_b64encode(url.encode()).decode()}'
```

and it should be reversible as well.

**Comment by candleindark** [comment]
> ```
> urlonly,options=url.split('?', 1) if '?' in url else url, ""
> path = '{quote(urlonly)}/?{base64.urlsafe_b64encode(url.encode()).decode()}'
> ```


There are two problems with this construction of path though. 

The first is that the length of `f"{base64.urlsafe_b64encode(url.encode()).decode()}"` alone is without bound.

For example, 

1.  When url is `http://www.example.com?oukqc=kocildl&jxema=hqesqcj&hljql=lfhmaqf&qpjte=knzfvnf&dobhw=hvtakpk&bvbki=henbjpk&azctd=mfaisdg&fahtn=vleumtg&zigxk=faqjume&cgxrz=xvssscs&vfdvi=pnlemtk&rumjq=ajmstxf&dwyci=nqgqepx&quzyk=lvevovx&gxzdp=nppdbzd&mavss=pnfzclx&eoiao=nfxdpjv&wxprr=fvppbfd&jshbg=jvcbwqr&myurv=ezrkgea`, its length is 404.
2. When url is `http://www.example.com/frywd/bfzfq/jvlhe/ytfmq/plwgq/epcpr/gmnvi/cjrwp/chjvh/etfoc/xfskh/wxevg/hyveg/wplhb/nltiy/rmvud/fobvt/liuit/oxibd/wcvvy/jupso/skwem/wtyxz/hlgyr/ryqlz/ncngz/quaop/fiwax/mjdoi/xfdfy/cmwsp/txsdx/mzabw/qwzoe/hrxcr/buelm/tbqsh/dkdir/uvhpe/kovhe/vyggj/sjfoc/sezka/rulcq/sdlbq/rlkqx/gjmza/kdqps/jpmhw/nrtkm/rovhy/ubywx/bnafk/flzzw/ijccw/bcqip/flxul/aagzw/yypqd/iwvea`, its length is 512.

So, this construction of path wouldn't solve our problem of having a long path in case of a long URL.

The second problem has to do with the first part of the construction, `f"{quote(urlonly)}"`. We will potentially have many URLs starting with `https://github.com/`. With this construction, we will have many subdirectories under the directory `https%3A//github.com/`. The number of these subdirectories may exceed the limitations allowed for the number entries in a directory.

I think at this point we can give up readability and reversibly for stability and simplicity. I am looking into a solution using hash functions, a solution similar to what git-annex uses for storing content of large files. I will propose a solution soon, but let me know if you think I heading the wrong direction.

**Comment by candleindark** [comment]
@yarikoptic Please consider the path of the dataset at a given URL as illustrated by the example below.

```python
from hashlib import blake2b
from base64 import b32encode

DIGEST_SIZE = 64

url = "http://www.example.com?hello=you&hey=world"

hashed_encoded_url = b32encode(
    blake2b(url.encode(), digest_size=DIGEST_SIZE, usedforsecurity=False).digest()
).decode()

path = f"{hashed_encoded_url[:2]}/{hashed_encoded_url[2:4]}/{hashed_encoded_url[4:]}"
```

`hashed_encoded_url` is guarantee to be of length 104 and `path` 106 regardless of the length of `url`. Additionally, because base-32 encoding, `b32encode`, is used instead of base-64, the generated paths work on case-insensitive filesystem as well (my Mac will be happy with that).

I chose BLAKE2 for hashing because it is supposed to be [fast](https://www.blake2.net/) and doesn't have known collision issues as MD5. Additionally, it even allows easy modification of [digest size](https://docs.python.org/3.9/library/hashlib.html#creating-hash-objects).

**Comment by yarikoptic** [comment]
> The number of these subdirectories may exceed the limitations allowed for the number entries in a directory.

- is there a limit?
- most likely would never will since the next level is user or organization. Some organizations, like `dandizarrs` could indeed have thousands of entries there.  But again - is there a limit an what is it?

> I am looking into a solution using hash functions,

I believe the idea behind using base64 was to have it reversible. Sure thing we could just take hash to guarantee stable length but then we cannot get back - figure out what url it is.  

re BLAKE2 -- if anything , we can go for SHA256 or even SHA1 . Even with md5 I think the concern of collisions is a moot point (md5 is cryptographically insecure in that you can cause collision if you want to but it is very unlikely to be observed in real life) given that we are looking just in some thousands, and unlikely few million of entries.  Here is what chatgpt says (not that it might wrong ;)):

>> chance of collision in md5 hash for 1 million items?
> MD5 is a 128-bit hash function that produces a fixed-size output for any input data. The probability of a collision in a hash function depends on the number of possible output values and the number of input values.
> In the case of MD5, there are 2^128 possible output values, which means that the probability of a collision for any two randomly chosen inputs is extremely low, about 1 in 2^128.
> For one million items, the number of possible pairs of items is given by (1,000,000 choose 2) = 499,999,500, which is a large number but still significantly smaller than 2^128.
> However, it is important to note that MD5 is no longer considered a secure hash function due to the existence of known attacks. Therefore, it is not recommended to use MD5 for any security-related applications.

Overall: 
- ATM I do not expect some ridiculously long (so there is no mapping on my file system in the way I suggested) or way too many URLs.  So I do not think your concerns would apply until someone proves me wrong with some ridiculously long URL
  - BTW we might need to sanitize/split more, e.g. on `::` as well . We can see URLs like `datalad-annex::file://{export}?type=directory&directory={{path}}&encryption=none&dladotgit=uncompressed` (replace placeholders with some sample values), and fsspec allows for other abominations like `libarchive://deeply/nested/path::ftp:///archive.7z` .
- let's retain reverse mappability (so hashes aren't the legit choice)

but then, may be, we should just keep it as is for now (base64), have a little helper to unmap back when really needed, and just add those few initial levels like we have in git/git-annex/you listed above. 

When/if we hit any limits, then we could consider some other solution.

**Comment by candleindark** [comment]
> is there a limit?

I assumed that there is a limit since git-annex's design takes that into consideration.

> most likely would never will since the next level is user or organization. Some organizations, like dandizarrs could indeed have thousands of entries there. But again - is there a limit an what is it?

The limit is different for different file systems. According to https://stackoverflow.com/questions/466521/how-many-files-can-i-put-in-a-directory/466596#466596, it seems that we should be well within any limit imposed by any file system, just as you expected. The same page also points out that there can be performance issues for a directories with many entries, but I think we will be unlikely to hit any limitation imposed by performance.

>I believe the idea behind using base64 was to have it reversible. Sure thing we could just take hash to guarantee stable length but then we cannot get back - figure out what url it is.

Why is it important to regenerate the URL from the cache path of the cloned dataset when we can just call `datalad siblings` or  `git remote show origin` in the cloned dataset?

> but then, may be, we should just keep it as is for now (base64), have a little helper to unmap back when really needed, and just add those few initial levels like we have in git/git-annex/you listed above (e.g. based on first characters of md5sum of the base64 encoded url).

Sure, we can do it that way. The following definition of path should exhibit that idea. Please let me know if it doesn't.

```python
from hashlib import md5
from base64 import urlsafe_b64encode

url = "http://www.github.com/some-org/some-repo?a=1&b=2"

b64encoded_url = urlsafe_b64encode(url.encode()).decode()
b64encoded_hashed_url = urlsafe_b64encode(
    md5(b64encoded_url.encode(), usedforsecurity=False).digest()
).decode()

path = f"{b64encoded_hashed_url[:2]}/{b64encoded_hashed_url[2:4]}/{b64encoded_url}"

print(path)

>>> E5/NO/aHR0cDovL3d3dy5naXRodWIuY29tL3NvbWUtb3JnL3NvbWUtcmVwbz9hPTEmYj0y
```

>BTW we might need to sanitize/split more, e.g. on :: as well . We can see URLs like datalad-annex::file://{export}?type=directory&directory={{path}}&encryption=none&dladotgit=uncompressed (replace placeholders with some sample values), and fsspec allows for other abominations like libarchive://deeply/nested/path::ftp:///archive.7z .

Issue https://github.com/datalad/datalad-registry/issues/147 is created to address this problem.

**Comment by yarikoptic** [comment]
> > I believe the idea behind using base64 was to have it reversible. Sure thing we could just take hash to guarantee stable length but then we cannot get back - figure out what url it is.
> 
> Why is it important to regenerate the URL from the cache path of the cloned dataset when we can just call `datalad siblings` or `git remote show origin` in the cloned dataset?

that is what I wondered too and so far came to the same conclusion as you did.  We do want urlencoded urls be passed within our API (http://github.com/datalad/datalad-registry/blob/HEAD/docs/openapi.yml) but then for them, after decoding we can indeed just compute hash for the location on the drive. 

So let's indeed switch to a simple digest, keep it md5 until we really run into conflict - so we might want to add in DB explicit column for the cached location, so we could at level of DB not only map back, but also quickly detect conflict if it arises (while adding a new URL, check if there is an entry with the same cached location).  But then that starts to beg question -- shouldn't we just mint a random UUID for that and be done?

if we were to go with `md5` -- imho could be as simple as 

```
In [29]: base64.b64encode(md5(url.encode(), usedforsecurity=False).digest()).decode()
Out[29]: 'P64xRjsPfDfqmm8qJ7Xf1Q=='
```
to base building the path of.

**Comment by candleindark** [comment]
> > > I believe the idea behind using base64 was to have it reversible. Sure thing we could just take hash to guarantee stable length but then we cannot get back - figure out what url it is.
> > 
> > 
> > Why is it important to regenerate the URL from the cache path of the cloned dataset when we can just call `datalad siblings` or `git remote show origin` in the cloned dataset?
> 
> that is what I wondered too and so far came to the same conclusion as you did. We do want urlencoded urls be passed within our API ([http://github.com/datalad/datalad-registry/blob/HEAD/docs/openapi.yml](https://github.com/datalad/datalad-registry/blob/HEAD/docs/openapi.yml)) but then for them, after decoding we can indeed just compute hash for the location on the drive.

In API V2, we will not need to encode a dataset URL to pass it in a request's URL. The dataset URL, unencoded, will be passed in the body of the request in JSON. In fact, the new API will accept a list of dataset URLs in JSON. So, we don't even needed to deal with base64 encoded URLs in the API level in the future.
 
> So let's indeed switch to a simple digest, keep it md5 until we really run into conflict - so we might want to add in DB explicit column for the cached location, so we could at level of DB not only map back, but also quickly detect conflict if it arises (while adding a new URL, check if there is an entry with the same cached location). But then that starts to beg question -- shouldn't we just mint a random UUID for that and be done?

The advantage of using a hash digest instead of a UUID is that we can at always produce the cache path from a URL. If we use UUID, then we will need to ask the database each time we need the cache path, potentially putting more load on the database. With that said, it is hard for me to think of a situation in which the app needs the cache path for a dataset URL yet not already has asked the database for the dataset URL's row, which contains the cache path. Let me know if you change your mind and decide to use UUID for the cache path instead.

> 
> if we were to go with `md5` -- imho could be as simple as
> 
> ```
> In [29]: base64.b64encode(md5(url.encode(), usedforsecurity=False).digest()).decode()
> Out[29]: 'P64xRjsPfDfqmm8qJ7Xf1Q=='
> ```
> 
> to base building the path of.

The following code should exhibit the current idea of the cache path. Let me know if it doesn't.

```python
from base64 import urlsafe_b64encode
from hashlib import md5

url = "http://www.github.com/some-org/some-repo?a=1&b=2"

b64encoded_url_md5 = urlsafe_b64encode(
    md5(url.encode(), usedforsecurity=False).digest()
).decode()

path = f"{b64encoded_url_md5[:2]}/{b64encoded_url_md5[2:4]}/{b64encoded_url_md5[4:]}"
```
Please note I used `base64.urlsafe_b64encode` instead of `base64.b64encode` because its result is [filesystem-safe as well as URL-safe](https://docs.python.org/3/library/base64.html#base64.urlsafe_b64encode).

The last thing I want to mention again is that base64 encoding is not safe for case-insensitive filesystems because its alphabet contains both upper and lower case letters. If we are to support any case-insensitive filesystem we should use base32, for example.

**Comment by candleindark** [comment]
This issue is closed with #159.
